include/async/schedulers/time_scheduler.hpp

```diff
@@ -0,0 +1,146 @@
+#pragma once
+
+#include <async/concepts.hpp>
+#include <async/env.hpp>
+#include <async/schedulers/timer_manager_interface.hpp>
+#include <async/tags.hpp>
+#include <async/type_traits.hpp>
+
+#include <stdx/concepts.hpp>
+
+#include <memory>
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+namespace async {
+namespace timer_mgr {
+template <typename Rcvr, typename Task> struct op_state_base : Task {
+    template <stdx::same_as_unqualified<Rcvr> R>
+    // NOLINTNEXTLINE(bugprone-forwarding-reference-overload)
+    constexpr explicit(true) op_state_base(R &&r) : rcvr{std::forward<R>(r)} {}
+
+    auto run() -> void final { set_value(std::move(rcvr)); }
+
+    [[no_unique_address]] Rcvr rcvr;
+};
+
+template <typename Duration, typename Rcvr, typename Task> struct op_state;
+
+template <typename Duration, typename Rcvr, typename Task>
+    requires unstoppable_token<stop_token_of_t<env_of_t<Rcvr>>>
+struct op_state<Duration, Rcvr, Task> final : op_state_base<Rcvr, Task> {
+    template <stdx::same_as_unqualified<Rcvr> R>
+    // NOLINTNEXTLINE(bugprone-forwarding-reference-overload)
+    constexpr explicit(true) op_state(R &&r, Duration dur)
+        : op_state_base<Rcvr, Task>{std::forward<R>(r)}, d{dur} {}
+
+    [[no_unique_address]] Duration d{};
+
+  private:
+    template <stdx::same_as_unqualified<op_state> O>
+    friend constexpr auto tag_invoke(start_t, O &&o) -> void {
+        detail::run_after(o, o.d);
+    }
+};
+
+template <typename Duration, typename Rcvr, typename Task>
+    requires(not unstoppable_token<stop_token_of_t<env_of_t<Rcvr>>>)
+struct op_state<Duration, Rcvr, Task> final : op_state_base<Rcvr, Task> {
+    template <stdx::same_as_unqualified<Rcvr> R>
+    // NOLINTNEXTLINE(bugprone-forwarding-reference-overload)
+    constexpr explicit(true) op_state(R &&r, Duration dur)
+        : op_state_base<Rcvr, Task>{std::forward<R>(r)}, d{dur} {}
+
+    [[no_unique_address]] Duration d{};
+
+  private:
+    struct stop_callback_fn {
+        auto operator()() -> void {
+            detail::cancel(*ops);
+            set_stopped(std::move(ops->rcvr));
+        }
+        op_state *ops;
+    };
+
+    template <stdx::same_as_unqualified<op_state> O>
+    friend constexpr auto tag_invoke(start_t, O &&o) -> void {
+        auto token = get_stop_token(get_env(o.rcvr));
+        if (token.stop_requested()) {
+            set_stopped(std::forward<O>(o).rcvr);
+        } else {
+            o.stop_cb.emplace(token, stop_callback_fn{std::addressof(o)});
```

Comment:
question: Do we have a race condition here it a stop runs at a higher priority? We have two scenarios I think:

1. Condition is checked on 69 and stop isn't requested. Before 72 sets the callback, a stop is requested. The timer runs the full course and no stop happens. This might be ok if the actual run checks stopped before actually running. Otherwise you might race the stop token, wait T time (lets say something really long like a second), and then actually run the action (pass value to the receiver).

2. The optional is mid-emplace and because it isn't properly constructed the stop callback function or op state pointer are torn and we execute bad code.

Reply:
I think lines 72 & 73 need to be swapped.

Reply:
That certainly takes care of part of the concern.

Reply:
Note: constructing a stop callback is thread-safe and will call the callback if stop has already been requested. Hence swapping 72 & 73 makes this safe.

Reply:
OK, I think there is no race, given the ordering of (now) lines 73 & 74, and the cancel check on line 60. Of course it is possible to request a stop "too late" but we should now get one of `set_value` or `set_stopped` called exactly once.

Reply:
I suppose the check on line 70 is really just an early out and also protecting against an immediate cancellation of work scheduled at zero offset.